Units prefix improvements

hsorby opened the pull request:
Looking at avoiding the exception raised by `std::stoi` this causes problems for web assembly binaries.  Hopefully other exceptions are harder to trigger.

Addresses #845.

hsorby committed:
+1 is a valid CellML integer.

hsorby committed:
Avoid causing exception when parsing prefix.  Make use of a utility funciton to convert a prefix to an int.

nickerso commented on src/units.cpp:
```diff
@@ -328,14 +333,26 @@ void Units::addUnit(const std::string &reference, const std::string &prefix, dou
-        int prefixInteger = std::stoi(prefix);
-        if (prefixInteger != 0.0) {
+
+    bool predefinedPrefix = false;
```
can't this use the new `convertPrefixToInt()` method?

hsorby replied:
No because we don't know the state of the prefix in this method.  The `convertPrefixToInt()` method is for when the input is well formed.  That is, it is either a known prefix string or a valid CellML integer.

nickerso reviewed (dismissed):
looks good once the clang formatting is kept happy...

hsorby committed:
Make clang format happy in units.cpp.

agarny commented on src/units.cpp:
```diff
@@ -181,12 +181,17 @@ bool updateUnitMultiplier(const UnitsPtr &units, int direction, double &multipli
+                prefixMult = prefixInt;
             } else {
                 return false;
             }
```
I would use our new `convertPrefixToInt()` method for this, which means that `convertPrefixToInt()` would have to be reworked a bit to return `true` or `false` depending on whether the conversion was successful or not. IOW, I don't like "duplication" of code.

hsorby replied:
9a0c7b9

agarny reviewed (changes requested):
Looks good besides some minor refactoring that I would like to see done.

hsorby committed:
Refactor to use convertPrefixToInt().

hsorby committed:
Merge branch 'units-validation' of github.com:hsorby/libcellml into units-validation

agarny commented on src/units.cpp:
```diff
@@ -327,15 +329,27 @@ void Units::addUnit(const std::string &reference, const std::string &prefix, dou
     UnitDefinition ud;
     ud.mReference = reference;
+
     // Allow all nonzero user-specified prefixes
```
Need a full stop.

hsorby replied:
90d15a2

agarny commented on src/units.cpp:
```diff
@@ -185,8 +185,10 @@ bool updateUnitMultiplier(const UnitsPtr &units, int direction, double &multipli
+                prefixMult = convertPrefixToInt(pre);
             } else {
                 return false;
             }
```
I was expecting something like (not tested):
```c++
bool ok;

prefixMult = convertPrefixToInt(pre, &ok);

if (!ok) {
    return false;
}
```
with
```c++
int convertPrefixToInt(const std::string &in, bool *ok)
{
    int prefixInt = 0;

    if (ok != nullptr) {
        *ok = true;
    }

    if (isStandardPrefixName(in)) {
        prefixInt = standardPrefixList.at(in);
    } else if (in.empty()) {
        prefixInt = 0;
    } else if (isCellMLInteger(in)) {
        convertToInt(in, prefixInt);
    } else if (ok != nullptr) {
        *ok = false;
    }

    return prefixInt;
}
```

hsorby replied:
My preference is to keep `convertPrefixToInt` as it is.  What do you think @nickerso?

agarny replied:
You can still call `convertPrefixToInt()` the way you were doing it before, it's just that now you don't have to do those ugly tests before calling it in `updateUnitMultiplier()`. This makes code maintenance unnecessarily "complex".

hsorby replied:
True, but I am also looking at the other locations that this function is used and that proposal doesn't work well for those cases.

nickerso replied:
I've already approved this one twice, so I think its pretty clear I'm happy with it as is....

agarny replied:
What other locations? Re `convertPrefixToInt()`, if you want the exact same behaviour as your version, then you could have:
```c++
int convertPrefixToInt(const std::string &in, bool *ok)
{
    int prefixInt = 0;

    if (ok != nullptr) {
        *ok = true;
    }

    if (isStandardPrefixName(in)) {
        prefixInt = standardPrefixList.at(in);
    } else if (in.empty()) {
        prefixInt = 0;
    } else if ((ok == nullptr) || isCellMLInteger(in)) {
        convertToInt(in, prefixInt);
    } else if (ok != nullptr) {
        *ok = false;
    }

    return prefixInt;
}
```

hsorby replied:
`convertPrefixToInt` is used in the analyser and validator right.

agarny replied:
Indeed, they use it and I have just tested my implementation at https://github.com/cellml/libcellml/pull/846#discussion_r618845525 and the one at https://github.com/cellml/libcellml/pull/846#discussion_r618896294, and it all works fine. So, yes, I would use my implementation at https://github.com/cellml/libcellml/pull/846#discussion_r618845525.

agarny replied:
FWIW `convertPrefixToInt()` can be optimised a bit (since we initialise `prefixInt` to `0`:
```c++
int convertPrefixToInt(const std::string &in, bool *ok)
{
    int prefixInt = 0;

    if (ok != nullptr) {
        *ok = true;
    }

    if (isStandardPrefixName(in)) {
        prefixInt = standardPrefixList.at(in);
    } else if (isCellMLInteger(in)) {
        convertToInt(in, prefixInt);
    } else if ((ok != nullptr) && !in.empty()) {
        *ok = false;
    }

    return prefixInt;
}
```

hsorby replied:
7ebbfc9

agarny reviewed (changes requested):
I would go one extra step on that refactoring.

hsorby committed:
Add fullstop.

hsorby committed:
Remove code that is not needed from Units::addUnit.

hsorby committed:
Consolidate code for convertPrefixToInt

hsorby committed:
Merge branch 'main' into units-validation

nickerso committed:
Create utilities.h

Correcting typos and errors in the documentation for the convertPrefixToInt() method.

nickerso reviewed (approved):
Not a fan of the new `convertPrefixToInt()` method as it is now inconsistent with `convertToInt()` and `convertToReal()`. But otherwise this is good to go (assuming my changes to the documentation of that method are correct).

agarny commented:
> Not a fan of the new `convertPrefixToInt()` method as it is now inconsistent with `convertToInt()` and `convertToReal()`.

Those are internal methods and the new `convertPrefixToInt()` method ensures easier maintenance of our code. If we really want, we could update `convertToInt()` and `convertToReal()` to also take an optional `ok` argument.

> But otherwise this is good to go (assuming my changes to the documentation of that method are correct).

Looks good to me, so approving this PR, but will leave it to @hsorby to merge it in after he has confirmed he's also happy with your change.
